File: 0101-symmetric-tree/0101-symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sym(self, r1: Optional[TreeNode], r2: Optional[TreeNode]) -> bool:
        if (r1 is None) and (r2 is None):
            return True
        if (r1 is None) or (r2 is None):
            return False
        if r1.val != r2.val:
            return False
        return self.sym(r1.left, r2.right) and self.sym(r1.right, r2.left)

    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True
        return self.sym(root.left, root.right)

File: 0101-symmetric-tree/0101-symmetric-tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn is_symmetric(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        let res: bool = match root {
            Some(node) => Solution::sym(node.borrow().left.clone(), node.borrow().right.clone()),
            _ => true,
        };
        return res;
    }
    pub fn check_sym_with_val(n1: Rc<RefCell<TreeNode>>, n2: Rc<RefCell<TreeNode>>) -> bool {
        if (n1.borrow().val != n2.borrow().val) {
            return false;
        }
        return (Solution::sym(n1.borrow().left.clone(), n2.borrow().right.clone()) && Solution::sym(n1.borrow().right.clone(), n2.borrow().left.clone()));
    }
    pub fn sym(r1: Option<Rc<RefCell<TreeNode>>>, r2: Option<Rc<RefCell<TreeNode>>>) -> bool {
        let res: bool = match (r1, r2) {
            (Some(n1), Some(n2)) => Solution::check_sym_with_val(n1, n2),
            (Some(n1), _) => false,
            (_, Some(n2)) => false,
            (_, _) => true,
        };
        return res;
    }
}
